File: PROGRAMMERS/LEVEL_02/Q17684.cpp
//https://programmers.co.kr/learn/courses/30/lessons/17684
//압축
#include <iostream>

#include <string>
#include <vector>
#include <algorithm>

using namespace std;

vector<int> solution(string msg) {

    // 사전 초기화 (알파벳만 포함)
    const int MAX = 26;
    vector<string> dic(MAX + 1);
    for (int i = 0; i < MAX; ++i)
    {
        dic[i + 1] = i + 'A';
    }

    // LZW 압축
    vector<int> answer;
    int strt = 0, len = msg.size();
    while(strt < len)
    {
        int end = strt;
        string now, next;
        vector<string>::iterator it_now, it_next = dic.begin();

        // 현재 위치에서 사전에 등록된 가장 긴 단어와, 그 다음단어 찾기
        while (end <= len)
        {
            ++end;
            next = msg.substr(strt, end - strt);
            it_next = find(dic.begin(), dic.end(), next);

            // 찾은 단어가 사전에 등록되어 있으면 now로 교환하고 다음 단어 찾기
            if (it_next != dic.end())
            {
                now = next;
                it_now = it_next;
            }
            else
            {
                --end;
                break;
            }
        }

        // answer벡터에 색인번호 추가
        answer.push_back(it_now - dic.begin());
        // 새로운 단어 사전에 추가
        dic.push_back(next);
        // 다음 단어 시작은 end번째 글자부터 
        strt = end;
    }

    return answer;
}

int main()
{
    /*
    string msg = "TOBEORNOTTOBEORTOBEORNOT";
    string msg = "KAKAO";
    */
    string msg = "ABABABABABABABAB";

    vector<int> rtn = solution(msg);
    for (int i : rtn)
    {
        cout << i << ' ';
    }

    return 0;
}

File: PROGRAMMERS/LEVEL_02/Q17684.java
// https://programmers.co.kr/learn/courses/30/lessons/17684?language=java
// [3차] 압축

package PROGRAMMERS.LEVEL_02;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Q17684 {

    public static void main(String[] args) {
        /*
        msg	                        answer
        KAKAO	                    [11, 1, 27, 15]
        TOBEORNOTTOBEORTOBEORNOT	[20, 15, 2, 5, 15, 18, 14, 15, 20, 27, 29, 31, 36, 30, 32, 34]
        ABABABABABABABAB	        [1, 2, 27, 29, 28, 31, 30]
         */
        String msg = "TOBEORNOTTOBEORTOBEORNOT";
        List<Integer> res = solution(msg);
        for (int i : res) {
            System.out.println(i);
        }
    }

    public static List<Integer> solution(String msg) {

        // initializing dictionary
        Map<String, Integer> dict = new HashMap<>();
        int idx = 0;
        while (idx++ <= 'Z' - 'A') {
            dict.put(Character.toString((char)(idx - 1 + 'A')) , idx);
        }

        int start = 0;
        int end = start + 1;
        int lastSearchIdx = 0;

        List<Integer> answerList = new ArrayList<>();

        while (end <= msg.length()) {

            String target = msg.substring(start, end);

            // when target exist : memorize idx -> check longer word
            if (dict.containsKey(target)) {
                lastSearchIdx = dict.get(target);
                end++;
                continue;
            }

            // when target not exist : add memorized idx on answer list -> check next word
            answerList.add(lastSearchIdx);
            dict.put(target, idx++);
            start = end - 1;
        }
        answerList.add(lastSearchIdx);

        return answerList;
    }
}
